source/simulator/core.cc

```diff
@@ -796,6 +799,7 @@ namespace aspect
           case Parameters<dim>::NonlinearSolver::Kind::iterated_Advection_and_Newton_Stokes:
           case Parameters<dim>::NonlinearSolver::Kind::single_Advection_iterated_Newton_Stokes:
           case Parameters<dim>::NonlinearSolver::Kind::single_Advection_no_Stokes:
+          case Parameters<dim>::NonlinearSolver::Kind::no_Advection_adjoint_Stokes:
             return true;
```

Comment:
Agreed that this seems wrong, but if I don't set this to true I get a runtime error during the parameter_update:

An error occurred in line <1684> of file </var/folders/8z/hlb6vc015qjggytkxn84m6_c0000gn/T/heltai/spack-stage/spack-stage-dealii-9.3.0-zy7k3uwnakcqjvrajvacy5l4jrl7eaex/spack-src/source/lac/trilinos_sparse_matrix.cc> in function
    void dealii::TrilinosWrappers::SparseMatrix::add(const dealii::TrilinosWrappers::SparseMatrix::size_type, const dealii::TrilinosWrappers::SparseMatrix::size_type, const dealii::TrilinosWrappers::SparseMatrix::size_type *, const dealii::TrilinosScalar *, const bool, const bool)
The violated condition was: 
    ierr == 0
Additional information: 
    An error with error number 2 occurred while calling a Trilinos
    function

Looking through the core.cc file I noticed that composition coupling is only enabled if this is true and wondered if this is required in my parameter update. However, enabling these directly for when the adjoint_stokes solver scheme was selected didn't work. I'm not sure what the issue is here because I can't imagine that it's temperature-related. Any ideas?

Reply:
This looks like the error you'd get if you're trying to write into the matrix and the sparsity pattern has not allocated an entry for the position you're trying to write into. Is this function used in determining what the sparsity pattern of the matrix should be? If so, maybe do
```
  if (solver_scheme_solves_advection_equations() || (solve == adjoint solver))
```

Reply:
I'm not sure where the sparsity pattern of the matrix is determined. I can change things such that at every time it checks for solver_scheme_solves_advection_equations it also checks || (solve == adjoint solver) and that works. Is that what you were suggesting?

Reply:
Yes.

As for the sparsity pattern, let's let that one go. We've already made the decision to use the (comp0,comp0) block to solve for compositional fields. Though I would prefer if you took the correct block of the right hand side vector at least.